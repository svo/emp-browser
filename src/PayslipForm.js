import React, { Component } from 'react';
import { FormGroup, ControlLabel, FormControl } from 'react-bootstrap';
import PropTypes from 'prop-types';

class PayslipForm extends Component {

  static propTypes = {
    year: PropTypes.number.isRequired,
    month: PropTypes.number.isRequired
  }

  constructor(props) {
    super(props);

    var date = new Date(this.props.year, this.props.month);
    var locale = "en-au";
    var month = date.toLocaleString(locale, { month: "long" });

    this.state = {
      month: month.toUpperCase(),
      first_name: "",
      last_name: "",
      annual_salary: 0
    };

    this.handleFirstNameChange = this.handleFirstNameChange.bind(this);
    this.handleLastNameChange = this.handleLastNameChange.bind(this);
    this.handleAnnualSalaryChange = this.handleAnnualSalaryChange.bind(this);
  }

  validFirstName() {
    const length = this.state.first_name.length;
    return (length > 0) ? 'success' : 'error';
  }

  validLastName() {
    const length = this.state.last_name.length;
    return (length > 0) ? 'success' : 'error';
  }

  validAnnualSalary() {
    var value = parseFloat(this.state.annual_salary);

    if (typeof value === 'number' &&
      isFinite(value) &&
      Math.floor(value) === value &&
      value > 0) {

      return 'success';
    } else {
      return 'error';
    }
  }

  handleFirstNameChange(e) {
    this.setState({ first_name: e.target.value });
  }

  handleLastNameChange(e) {
    this.setState({ last_name: e.target.value });
  }

  handleAnnualSalaryChange(e) {
    this.setState({ annual_salary: e.target.value });
  }

  render() {
    return (
      <form>
        <FormGroup controlId="first_name" validationState={this.validFirstName()}>
          <ControlLabel>First Name</ControlLabel>
          <FormControl type="text"
            placeholder="Sean"
            value={this.state.first_name}
            onChange={this.handleFirstNameChange} />
          <FormControl.Feedback />
        </FormGroup>
        <FormGroup controlId="last_name" validationState={this.validLastName()}>
          <ControlLabel>Last Name</ControlLabel>
          <FormControl type="text"
            placeholder="Van Osselaer"
            value={this.state.last_name}
            onChange={this.handleLastNameChange} />
          <FormControl.Feedback />
        </FormGroup>
        <FormGroup controlId="annual_salary" validationState={this.validAnnualSalary()}>
          <ControlLabel>Annual Salary</ControlLabel>
          <FormControl type="number"
            placeholder="175000"
            value={this.state.annual_salary}
            onChange={this.handleAnnualSalaryChange} />
          <FormControl.Feedback />
        </FormGroup>
        <FormGroup controlId="year">
          <ControlLabel>Year</ControlLabel>
          <FormControl type="text"
                       defaultValue={this.props.year}
                       placeholder={this.props.year} />
        </FormGroup>
        <FormGroup controlId="month">
          <ControlLabel>Month</ControlLabel>
          <FormControl componentClass="select" placeholder="..."
              defaultValue={this.state.month}>
            <option value="JANUARY">January</option>
            <option value="FEBRUARY">February</option>
            <option value="MARCH">March</option>
            <option value="APRIL">April</option>
            <option value="MAY">May</option>
            <option value="JUNE">June</option>
            <option value="JULY">July</option>
            <option value="AUGUST">August</option>
            <option value="SEPTEMBER">September</option>
            <option value="OCTOBER">October</option>
            <option value="NOVEMBER">November</option>
            <option value="DECEMBER">December</option>
          </FormControl>
        </FormGroup>
        <FormGroup controlId="super_rate">
          <ControlLabel>Super Rate</ControlLabel>
          <FormControl type="text" placeholder="..."
            defaultValue="9"/>
        </FormGroup>
      </form>
    );
  }
}

export default PayslipForm;
