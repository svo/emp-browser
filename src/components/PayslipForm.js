import React, { Component } from 'react';
import { FormGroup, ControlLabel, FormControl, Button, ButtonGroup } from 'react-bootstrap';
import PropTypes from 'prop-types';
import format from 'date-format';
import PayslipActions from '../actions/PayslipActions';
import PayslipStore from '../stores/PayslipStore';

class PayslipForm extends Component {

  static propTypes = {
    now: PropTypes.object.isRequired
  }

  constructor(props) {
    super(props);

    this.state = {
      first_name: "",
      last_name: "",
      annual_salary: 0,
      month: this.props.now,
      super_rate: 9
    };

    this.handleFirstNameChange = this.handleFirstNameChange.bind(this);
    this.handleLastNameChange = this.handleLastNameChange.bind(this);
    this.handleAnnualSalaryChange = this.handleAnnualSalaryChange.bind(this);
    this.handleMonthChange = this.handleMonthChange.bind(this);
    this.handleSuperRateChange = this.handleSuperRateChange.bind(this);
    this.handleCreateClick = this.handleCreateClick.bind(this);
    this.onChange = this.onChange.bind(this);
  }

  getInitialState() {
    return PayslipStore.getState();
  }

  componentDidMount() {
    PayslipStore.listen(this.onChange);
  }

  componentWillUnmount() {
    PayslipStore.unlisten(this.onChange);
  }

  onChange(state) {
    this.setState({location: state.payslip.location});
  }

  validFirstName() {
    const length = this.state.first_name.length;
    return (length > 0) ? 'success' : 'error';
  }

  validLastName() {
    const length = this.state.last_name.length;
    return (length > 0) ? 'success' : 'error';
  }

  validAnnualSalary() {
    var value = parseFloat(this.state.annual_salary);

    if (typeof value === 'number' &&
      isFinite(value) &&
      Math.floor(value) === value &&
      value > 0) {

      return 'success';
    } else {
      return 'error';
    }
  }

  validSuperRate() {
    var value = parseFloat(this.state.super_rate);

    if (typeof value === 'number' &&
      isFinite(value) &&
      value >= 0 &&
      value <= 50) {

      return 'success';
    } else {
      return 'error';
    }
  }

  valid() {
    return this.validSuperRate() === 'success' &&
      this.validAnnualSalary() === 'success' &&
      this.validFirstName() === 'success' &&
      this.validLastName() === 'success';
  }

  handleFirstNameChange(e) {
    this.setState({ first_name: e.target.value });
  }

  handleLastNameChange(e) {
    this.setState({ last_name: e.target.value });
  }

  handleAnnualSalaryChange(e) {
    this.setState({ annual_salary: e.target.value });
  }

  handleMonthChange(e) {
    this.setState({ month: new Date(e.target.value) });
  }

  handleSuperRateChange(e) {
    this.setState({ super_rate: e.target.value });
  }

  handleCreateClick() {
    PayslipActions.create(this.state);
  }

  render() {
    return (
      <form>
        <FormGroup controlId="first_name" validationState={this.validFirstName()}>
          <ControlLabel>First Name</ControlLabel>
          <FormControl type="text"
            placeholder="Sean"
            value={this.state.first_name}
            onChange={this.handleFirstNameChange} />
          <FormControl.Feedback />
        </FormGroup>
        <FormGroup controlId="last_name" validationState={this.validLastName()}>
          <ControlLabel>Last Name</ControlLabel>
          <FormControl type="text"
            placeholder="Van Osselaer"
            value={this.state.last_name}
            onChange={this.handleLastNameChange} />
          <FormControl.Feedback />
        </FormGroup>
        <FormGroup controlId="annual_salary" validationState={this.validAnnualSalary()}>
          <ControlLabel>Annual Salary</ControlLabel>
          <FormControl type="number"
            placeholder="175000"
            value={this.state.annual_salary}
            onChange={this.handleAnnualSalaryChange} />
          <FormControl.Feedback />
        </FormGroup>
        <FormGroup controlId="year">
          <ControlLabel>Year</ControlLabel>
          <FormControl type="month"
            value={format('yyyy-MM', this.state.month)}
            onChange={this.handleMonthChange} />
        </FormGroup>
        <FormGroup controlId="super_rate" validationState={this.validSuperRate()}>
          <ControlLabel>Super Rate</ControlLabel>
          <FormControl type="number"
            placeholder="9"
            value={this.state.super_rate}
            onChange={this.handleSuperRateChange} />
          <FormControl.Feedback />
        </FormGroup>
        <ButtonGroup>
          <Button bsStyle="primary" onClick={this.handleCreateClick}>
            Create
          </Button>
          <Button bsStyle="success" href={this.state.location}
            disabled={this.state.location === undefined}>
            Download
          </Button>
        </ButtonGroup>
      </form>
    );
  }
}

export default PayslipForm;
