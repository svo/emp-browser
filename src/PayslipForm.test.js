import React from 'react';
import ReactDOM from 'react-dom';
import PayslipForm from './PayslipForm';
import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const tree = renderer.create(
    <PayslipForm year={2017} month={1} />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

describe("properties", function() {
  it('should have correct year property', () => {
    const div = document.createElement('div');
    var payslipForm = ReactDOM.render(<PayslipForm year={1976} month={3} />, div);
    expect(payslipForm.props.year).toEqual(1976);
  });

  it('should have correct month property', () => {
    const div = document.createElement('div');
    var payslipForm = ReactDOM.render(<PayslipForm year={1976} month={3} />, div);
    expect(payslipForm.props.month).toEqual(3);
  });

});

describe("state", function() {
  it('should have correct month state', () => {
    const div = document.createElement('div');
    var payslipForm = ReactDOM.render(<PayslipForm year={1976} month={3} />, div);
    expect(payslipForm.state.month).toEqual("APRIL");
  });
});

describe("validation", function() {
  describe("of first name", function() {
    it('should report invalid first name', () => {
      const div = document.createElement('div');
      var payslipForm = ReactDOM.render(<PayslipForm year={1976} month={3} />, div);

      expect(payslipForm.validFirstName()).toEqual('error')
    });

    it('should report valid first name', () => {
      const div = document.createElement('div');
      var payslipForm = ReactDOM.render(<PayslipForm year={1976} month={3} />, div);

      payslipForm.setState({first_name: "Sean"});

      expect(payslipForm.validFirstName()).toEqual('success')
    });
  });

  describe("of last name", function() {
    it('should report invalid last name', () => {
      const div = document.createElement('div');
      var payslipForm = ReactDOM.render(<PayslipForm year={1976} month={3} />, div);

      expect(payslipForm.validLastName()).toEqual('error')
    });

    it('should report valid last name', () => {
      const div = document.createElement('div');
      var payslipForm = ReactDOM.render(<PayslipForm year={1976} month={3} />, div);

      payslipForm.setState({last_name: "Van Osselaer"});

      expect(payslipForm.validLastName()).toEqual('success')
    });
  });

  describe("of annual salary", function() {
    it('should report invalid annual salary if not provided', () => {
      const div = document.createElement('div');
      var payslipForm = ReactDOM.render(<PayslipForm year={1976} month={3} />, div);

      expect(payslipForm.validAnnualSalary()).toEqual('error')
    });

    it('should report invalid annual salary for decimal values', () => {
      const div = document.createElement('div');
      var payslipForm = ReactDOM.render(<PayslipForm year={1976} month={3} />, div);

      payslipForm.setState({annual_salary: "1.01"});

      expect(payslipForm.validAnnualSalary()).toEqual('error')
    });

    it('should report valid annual salary', () => {
      const div = document.createElement('div');
      var payslipForm = ReactDOM.render(<PayslipForm year={1976} month={3} />, div);

      payslipForm.setState({annual_salary: "1"});

      expect(payslipForm.validAnnualSalary()).toEqual('success')
    });
  });
});
