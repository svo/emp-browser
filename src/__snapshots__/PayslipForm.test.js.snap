// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<form>
  <div
    className="form-group has-feedback has-error"
  >
    <label
      className="control-label"
      htmlFor="first_name"
    >
      First Name
    </label>
    <input
      className="form-control"
      id="first_name"
      onChange={[Function]}
      placeholder="Sean"
      type="text"
      value=""
    />
    <span
      className="form-control-feedback glyphicon glyphicon-remove"
    />
  </div>
  <div
    className="form-group has-feedback has-error"
  >
    <label
      className="control-label"
      htmlFor="last_name"
    >
      Last Name
    </label>
    <input
      className="form-control"
      id="last_name"
      onChange={[Function]}
      placeholder="Van Osselaer"
      type="text"
      value=""
    />
    <span
      className="form-control-feedback glyphicon glyphicon-remove"
    />
  </div>
  <div
    className="form-group has-feedback has-error"
  >
    <label
      className="control-label"
      htmlFor="annual_salary"
    >
      Annual Salary
    </label>
    <input
      className="form-control"
      id="annual_salary"
      onChange={[Function]}
      placeholder="175000"
      type="number"
      value={0}
    />
    <span
      className="form-control-feedback glyphicon glyphicon-remove"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className="control-label"
      htmlFor="year"
    >
      Year
    </label>
    <input
      className="form-control"
      defaultValue={2017}
      id="year"
      placeholder={2017}
      type="text"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className="control-label"
      htmlFor="month"
    >
      Month
    </label>
    <select
      className="form-control"
      defaultValue="FEBRUARY"
      id="month"
      placeholder="..."
      type={undefined}
    >
      <option
        value="JANUARY"
      >
        January
      </option>
      <option
        value="FEBRUARY"
      >
        February
      </option>
      <option
        value="MARCH"
      >
        March
      </option>
      <option
        value="APRIL"
      >
        April
      </option>
      <option
        value="MAY"
      >
        May
      </option>
      <option
        value="JUNE"
      >
        June
      </option>
      <option
        value="JULY"
      >
        July
      </option>
      <option
        value="AUGUST"
      >
        August
      </option>
      <option
        value="SEPTEMBER"
      >
        September
      </option>
      <option
        value="OCTOBER"
      >
        October
      </option>
      <option
        value="NOVEMBER"
      >
        November
      </option>
      <option
        value="DECEMBER"
      >
        December
      </option>
    </select>
  </div>
  <div
    className="form-group"
  >
    <label
      className="control-label"
      htmlFor="super_rate"
    >
      Super Rate
    </label>
    <input
      className="form-control"
      defaultValue="9"
      id="super_rate"
      placeholder="..."
      type="text"
    />
  </div>
</form>
`;
